tun/tun_illumos.go

```diff
@@ -0,0 +1,355 @@
+//go:build illumos && amd64
+// +build illumos,amd64
+
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright (C) 2017-2021 WireGuard LLC. All Rights Reserved.
+ * Copyright 2019 Joyent, Inc.
+ */
+
+package tun
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	tunNode = "/dev/tun"
+)
+
+type NativeTun struct {
+	fd      int
+	name    string
+	tunFile *os.File
+	events  chan Event
+	errors  chan error
+}
+
+// XXX This is largely copied from the freebsd code...
+// XXX How do I test if it is actually working?
+func (tun *NativeTun) routineRouteListener() {
+	var (
+		statusUp  bool
+		statusMTU int
+	)
+
+	tunIfindex, err := func() (int, error) {
+		iface, err := net.InterfaceByName(tun.name)
+		if err != nil {
+			return -1, err
+		}
+		return iface.Index, nil
+	}()
+	if err != nil {
+		tun.errors <- err
+		return
+	}
+
+	defer close(tun.events)
+
+	routeSocket, err := unix.Socket(unix.AF_ROUTE, unix.SOCK_RAW, unix.AF_UNSPEC)
+	if err != nil {
+		tun.errors <- err
+		return
+	}
+	defer unix.Close(routeSocket)
+
+	data := make([]byte, os.Getpagesize())
+	for {
+	retry:
+		n, err := unix.Read(routeSocket, data)
+		if err != nil {
+			if errors.Is(err, unix.EINTR) {
+				goto retry
+			}
+			tun.errors <- err
+			return
+		}
+
+		if n < 14 {
+			continue
+		}
+
+		if data[3 /* type */] != unix.RTM_IFINFO {
+			continue
+		}
+		ifindex := int(*(*uint16)(unsafe.Pointer(&data[12 /* ifindex */])))
+		if ifindex != tunIfindex {
+			continue
+		}
+
+		iface, err := net.InterfaceByIndex(ifindex)
+		if err != nil {
+			//tun.errors <- err
+			//return
+			// XXX returning this error breaks everything??
```

Comment:
This suggests that the struct being parsed is not the same on illumos and freebsd, where you got this code from. That would result in finding nonsensical indices, and having the interface lookup fail despite the tunnel still existing. Looking at golang.org/x/net/route, which has parsers for these RIB messages, it looks like there's 2-3 variants that put the ifindex at different offsets.

Easy fix for this is to look up in the illumos source what offset the ifindex sits at, and adjust the unsafe cast above.

Longer term, it might be nice to teach x/net/route to parse illumos RIB messages, and switch wireguard-go to using those parsers for all OSes, to remote another use of unsafe from this codebase. I think that should be a separate refactoring PR that makes the change across all platforms though, for this particular PR I think it's better to stick to the existing parsing strategy, which all the other non-linux unices also use.